File: Golang/QuickSort.go
package quicksort

func sortColors(nums []int) {
	partition(nums, 0, len(nums)-1)
}

func partition(nums []int, start int, end int) {
	if start >= end {
		return
	}
	j := start - 1
	pivot := nums[end]
	for i := start; i <= end; i++ {
		if nums[i] < pivot {
			j++
			nums[i], nums[j] = nums[j], nums[i]
		}
	}
	nums[end], nums[j+1] = nums[j+1], nums[end]
	partition(nums, start, j)
	partition(nums, j+2, end)
}

File: Typescript/QuickSort.ts
function quickSort(nums: number[]): void {
	function partition(start: number, end: number) {
		if(start >= end) {
			return
		}
		let pivot = nums[end]
		let j = start - 1; // j is the last index where nums[j] < pivot
		for(let i = start; i <= end; i++) {
			if(nums[i] < pivot) {
				j++
				[nums[i], nums[j]] = [nums[j], nums[i]]
			}
		}
		j++
		[nums[j], nums[end]] = [nums[end], nums[j]]
		partition(start, j - 1)
		partition(j + 1, end)
	}
    return partition(0, nums.length - 1)
};
